### PR TITLE
Introduce licensed plugins

### PR DESCRIPTION
**Pull request by pugnascotia**
This PR introduces the concept of "licensed" plugins. Such plugins may only be installed on installations of the default distribution, and this is enforced by the plugin installer. This PR also moves the `quote-aware-fs` plugin to the `x-pack` directory, and marks it as licensed.

Note that I didn't move the plugin source under `x-pack/plugin` because all the existing x-pack plugins are actually bundles as modules into the default distribution, whereas the `quota-aware-fs` plugin needs to remain a standalone plugin.

Note that release-manager will also need updated to know about this new plugin location, so that the plugin can be build and included in the release. I will prepare another and intend to merge them at the same time.

### CODE REVIEW

**Review by jasontedor — APPROVED** [review]
I skimmed this and at a high level it looks good to me. I wonder if (in a follow up) we should rename x-pack/plugin to x-pack/modules and create x-pack/plugins and migrate this quota-aware plugin there?

Please do give time for Mark or Ryan to leave a detailed review though, I’ve only checked things at a high level, but not the nitty details.

**Comment by pugnascotia** [comment]
@rjernst / @mark-vieira could I get a review, please?

File: distribution/tools/plugin-cli/src/main/java/org/elasticsearch/plugins/LicensedPlugin.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+import org.elasticsearch.Build;
+import org.elasticsearch.cli.ExitCodes;
+import org.elasticsearch.cli.Terminal;
+import org.elasticsearch.cli.UserException;
+
+import java.util.List;
+
+public class LicensedPlugin {
+
+    static void checkCanInstallationProceed(Terminal terminal, Build.Flavor flavor, PluginInfo info) throws Exception {
```
Comment (rjernst) [inline distribution/tools/plugin-cli/src/main/java/org/elasticsearch/plugins/LicensedPlugin.java:31]: Is there a reason this method couldn't exist within InstallPluginCommand?

File: docs/build.gradle [hunk 2]
```diff
@@ -87,7 +87,7 @@ project.rootProject.subprojects.findAll { it.parent.path == ':plugins' }.each {
    * doesn't make sense to waste time installing them.
    * Also skip quota-aware-fs since it has to be configured in order to use
```
Comment (rjernst) [inline docs/build.gradle:88]: This comment is out of date?

**Review by rjernst — APPROVED** [review]
LGTM. I also agree with Jason's suggestion to rename x-pack/plugin to x-pack/modules (as a followup).

**Review by mark-vieira — APPROVED** [review]
In addition to the `licensed` flag, shouldn't these plugin bundles include a license in the zip as well? So far as I can tell we aren't doing that. I think if we are going to distribute Elastic licensed artifacts, those artifacts should contain a copy of our license in addition to be prompted on install.

**Comment by rjernst** [comment]
@mark-vieira This happens automatically to everything under `x-pack`. See `x-pack/build.gradle` which sets the license on any PluginBuildPlugin below it.

File: distribution/tools/plugin-cli/src/main/java/org/elasticsearch/plugins/LicensedPlugin.java (hunk 1 above)
Comment (pugnascotia) [reply distribution/tools/plugin-cli/src/main/java/org/elasticsearch/plugins/LicensedPlugin.java]: There used to be more code involved here, which is why I split it out, but as you rightly point out it's just short-ish method now. I've folded it back into the installer class.

File: docs/build.gradle (hunk 2 above)
Comment (pugnascotia) [reply docs/build.gradle]: I've removed the comment.

**Comment by pugnascotia** [comment]
I created https://github.com/elastic/elasticsearch/issues/65125 for rearranging the x-pack modules.

**Comment by pugnascotia** [comment]
@elasticmachine run elasticsearch-ci/2

**Comment by mark-vieira** [comment]
> @mark-vieira This happens automatically to everything under x-pack. See x-pack/build.gradle which sets the license on any PluginBuildPlugin below it.

Thanks. I was looking in `PluginBuildPlugin` itself for that.

**Pull request merged**